Minutes 2021-08-18 (#250)
solid/authorization-panel: meetings/2021-08-18.md

# 2021-08-18 Solid Authorization

https://meet.jit.si/solid-authorization

## Agenda

* Minutes 2021-08-11 - https://github.com/solid/authorization-panel/pull/246
* [PR 244: Effective ACR discovery](https://github.com/solid/authorization-panel/pull/244)
* [PR 249: Added first delegation use case](https://github.com/solid/authorization-panel/pull/249)
* [PR 241: Fixes and add story for privacy](https://github.com/solid/authorization-panel/pull/241)
* Access Modes currently used by: WAC, ACP (5min)
* Reliance on client interpreting IRIs (aka. `/` hierarchy): WAC, ACP (5min)

## Present

* Barath
* Elf
* Henry
* Matthieu
* Justin

## Minutes

### Effective ACR discovery

Henry: Issues related to optimizing the number of requests for effective ACR discovery:
- [Support Trig serialization of Access Control Resources](https://github.com/solid/authorization-panel/issues/247)
- [Define another relation for non-created ACR](https://github.com/solid/authorization-panel/issues/248)

> I describe my favorite option in issue 248, which adds a new "control" link type to link to the potential (not-yet-created) ACR for when a default ACR is being used.

Reviewer: ```suggestion
I describe my favorite option in issue #248, which adds a new "control" link type to link to the potential (not-yet-created) ACR for when a default ACR is being used.
```

I need to create these issues to be able to link to them in order [PR 244: Effective ACR discovery](https://github.com/solid/authorization-panel/pull/244), and so for it to be merged.

Matthieu: Given that we have 4 potential solutions to minimising the number of requests involved in discovering the effective ACR, is everyone ok to merge #244?

Henry: Yes.

Elf: +1

Justin: +1

Matthieu: Merged.

Henry: The last solution might be useful for ACP too, if you want to weaken the requirement that every resource must have one ACR.

Matthieu: I'm not sure that falls into the ACP model, but let's review once the draft ACP PR is ready for review (that will be done today).

### Delegation use case

> Henry: Could you give us a review of that use case Elf?

Reviewer: ```suggestion
Henry: Could you give us a review of that use case, Elf?
```

Elf: How do you feel about using the [GNAP terminology](https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-00.html#name-roles) of Resource owner, server...?

Justin: I like using that terminology because I think it's well thought out and aligns with where we're going in terms of different actors and roles and responsibilities.

Henry: Btw, a bit out of context, but how long is the GNAP spec since we mentioned it (I think it was 80 pages last time I looked)?

> Justin: Maybe Pavlik you could extract the relevant terminology.

Reviewer: ```suggestion
Justin: Maybe, Pavlik, you could extract the relevant terminology?
```

> Henry: It's fine, I need to study that more thoroughly. Justin, you maintain this document (authorization UCs) so you might know what is relevant best.

Reviewer: ```suggestion
Henry: It's fine, I need to study that more thoroughly. Justin, you maintain this document (authorization UCs), so you might know what is most relevant?
```

> Justin: The AC UCs probably don't have a huge delta, maybe Pavlik you could do a PR and see how it aligns. Maybe spend 15mn to do some side by side. My experience with GNAP is that it is going to provide more specificity about the actors.

Reviewer: ```suggestion
Justin: The AC UCs probably don't have a huge delta. Maybe, Pavlik, you could do a PR and see how it aligns? Maybe spend 15mn to do some side by side. My experience with GNAP is that it is going to provide more specificity about the actors.
```

> Elf: I also added a part about delegation vs. impersonation. Bob would be able to access, but from the Authorization Server's perspective, it would be Alice accessing (impersonation) with delegation, ACME systems would know that Bob is accessing. Let's take a week to process that.

Reviewer: ```suggestion
Elf: I also added a part about delegation vs. impersonation. Bob would be able to access, but from the Authorization Server's perspective, it would be Alice accessing (impersonation). With delegation, ACME systems would know that Bob is accessing. Let's take a week to process that.
```

### Fixes and add story for privacy

* [PR 241: Fixes and add story for privacy](https://github.com/solid/authorization-panel/pull/241)

Elf: I think the musings about clients and so on distract from the main goal of focusing on inheritance. It would be good to have end to end but maybe in a primer or something like that.

Henry: I'm happy to start moving stuff to another place at some point in the future. But, at present, we have only three UC evaluations. It is a lot of work to do each, and so I'd rather have one that has a large coverage to test the concepts, rather than start off doing another one.

These evaluations are, in my view, meant to show how the whole system is working.
Unless you have a client and a server working together, we don't have a protocol. We need a system where the client can discover what the control rules are.

> Elf: That is UC3, would you add those concerns to every use case?

Reviewer: It's OK for some Requirement(s) to be required by multiple Use Cases.  Typically, there will be some Requirements that are required by two or more Use Cases, and some that are required by only one Use Case.  Use Cases should be realistically simple (or complicated), and not leave out important elements of process just to limit the number of Requirements they have.  It's OK for one Use Case to have several Requirements.

Henry: if a client gets a 40x with no link to an ACL or a link but no access to the rules, then the client cannot know how to authenticate, and so that would be the end of the story as far as the protocol is concerned, yes.

Barath: Let's say I come and request for your sensitive files and get a flat no with no details about the rules. This seems like wanted behavior. Sensitive rules except when being an admin should not be visible.
It's a difference between the control plane and resource plane.

Henry: If you start with a client making a request on a resource, receiving a 40x with no more information, that's the end of it.
The client cannot know how to continue.

But that is not the only use case. For example, we need a way for administrators to add rules on who can access (read and write).

But we also have a use case for [access by payment](https://solid.github.io/authorization-panel/authorization-ucr/#conditional-payment) or access for [users over 18](https://solid.github.io/authorization-panel/authorization-ucr/#uc-minimalcredentials). 
For those use cases, there need be no problem with the access control rules being known publically.  

> In those cases a 401 with a link to the rules that enables a client to know what to present to gain access (eg proof of being over 18), is completely fine.

Reviewer: ```suggestion
In those cases, a `401` with a link to the rules that enable a client to know what to present to gain access (e.g., proof of being over 18) is completely fine.
```

Indeed, there are many real-life use-cases where it is widely accepted that access control rules should be public: 

>  * If you walk in the country side you may see a sign "Do not trespass" sign. 

Reviewer: ```suggestion
 * If you walk in the countryside, you may see a "Do not trespass" sign. 
```

 * A shop has prices on items, that can be considered as access control rules: you need to present a money token to be able to take them out
 *... 
All these rules are public. 

> Barath: I don't know that this analogy translates to my personal data. The very fact to reveal that something exists is problematic.

Reviewer: ```suggestion
Barath: I don't know that this analogy translates to my personal data. The very act of revealing that something exists is problematic.
```

I would say that people that are privacy conscious don't put their data in public.

> Henry: We need to keep things in perspective here. The largest social networks are currently seeing all the information about every one of their users. And we want to help fix that: we are aiming to make it possible to socialise without a panopticon architecture. If we do that

Reviewer: ```suggestion
Henry: We need to keep things in perspective here. The largest social networks are currently seeing all the information about every one of their users. And we want to help fix that: we are aiming to make it possible to socialise without a panopticon architecture. If we do that,
```

How are they going to know if they can have access?
This is not a problem: we can have an ACL that is public and that describes the members who have access as being those who belong to  a group - and yet that grop need only be readable to the members of the group. So the ACL can be public, without there being information leakage. (That is what is described [in the use case we are discussing](https://github.com/solid/authorization-panel/blob/main/proposals/evaluation/uc-3-inheritance.md#wac)).

Elf: Let's get to the queue.

Matthieu: I would be happy to have a specific use case for discovery and credentials selection.

Justin: I think we should try to focus the debate. It is important in the overall system that clients and agents have knowledge of what they have access to. There are different ways of addressing that question. What is responsible for giving you slices of that knowledge is a different question.
In WAC and ACP, what is the scope of those?
How do you get knowledge about what you have access to and how to gain access to it to not over-share your credentials.
My feeling is that it is not necessarily in terms of ACP or WAC that you would disclose required credentials.

> Can you access resource X, I look at it like a firewall. You either get there or not. You can infer things sometimes but a firewall will never say to you "Justin you're not allowed to access the HTTP port on that endpoint". Me knowing that I now have access to a system, may have come out of band or through some other protocol. We spend a lot of time in interop designing how to give clients this kind of information.

Reviewer: ```suggestion
Can you access resource X, I look at it like a firewall. You either get there or not. You can infer things sometimes but a firewall will never say to you, "Justin, you're not allowed to access the HTTP port on that endpoint". Me knowing that I now have access to a system may have come out of band or through some other protocol. We spend a lot of time in interop designing how to give clients this kind of information.
```

I think there is tension in whether this kind of information is in the scope of access control systems or in the scope of some other systems.

Barath: I agree.

Henry: I find that we're making excellent progress by looking at the client/server interaction in the various UCs.
I'm sorry if this does not quite fit with what interop is doing,... but we have been working here for a very long time on a way in which ACP/WAC can present this information, how it can be edited by clients, etc...
If the access control rule is not visible, then one might have to use a backchannel of course.
But Solid is not going to be successful if people have to make telephone calls and use backchannels all the time. Solid is going to be successful when it is a drag and drop process to edit an access control rule, and when clients can follow links to discover who has access in a well-defined manner that works across the web.

The reason why I'm trying to get this concern addressed in UC3, is to have a full story/protocol.

> Matthieu: I see Henry's solution as pretty good one. When public Access control resource (WAC or ACP) can be discovered by the client, then that's one way of discovering required credentials. In other cases Authorization Servers could communicate to interop system which credentials are needed to access the resource. That might need some standard interface for WAC/ACP systems to communicate with such interop systems.

Reviewer: ```suggestion
Matthieu: I see Henry's solution as pretty good one. When public Access control resource (WAC or ACP) can be discovered by the client, then that's one way of discovering required credentials. In other cases, Authorization Servers could communicate to interop system which credentials are needed to access the resource. That might need some standard interface for WAC/ACP systems to communicate with such interop systems.
```

> ...: We should also consider the case where the access rules are private, but maybe notification can be sent informing the relevant users of the resources they have access to and which credentials they need to access those resources that were shared with them. This seems like an out of band mechanism.

Reviewer: ```suggestion
...: We should also consider the case where the access rules are private, but maybe notification can be sent informing the relevant users of the resources they have access to, and which credentials they need to access those resources that were shared with them. This seems like an out of band mechanism.
```

I think we all agree that this is an important mechanism.
Maybe we can merge the UC3 PR now (keeping in mind we might separate credentials discovery concerns later on if a specific Use Case is formulated).

Elf: Ok to merge.

Henry: That's good. Note the UC3 example does make use
of such an out-of-band mechanism to get people into the private group. After all: the only way for one to know that one is a member is to be notified. But in that case we can easily see how LDN or a Linked-Data-based notification protocol can work.

### Closing

> Matthieu: We're a bit overtime, should we address the two outstanding points next week?

Reviewer: ```suggestion
Matthieu: We're a bit over time. Should we address the two outstanding points next week?
```

# 2021-08-18 Solid Authorization

https://meet.jit.si/solid-authorization

## Agenda

* Minutes 2021-08-11 - https://github.com/solid/authorization-panel/pull/246
* [PR 244: Effective ACR discovery](https://github.com/solid/authorization-panel/pull/244)
* [PR 249: Added first delegation use case](https://github.com/solid/authorization-panel/pull/249)
* [PR 241: Fixes and add story for privacy](https://github.com/solid/authorization-panel/pull/241)
* Access Modes currently used by: WAC, ACP (5min)
* Reliance on client interpreting IRIs (aka. `/` hierarchy): WAC, ACP (5min)

## Present

* Barath
* Elf
* Henry
* Matthieu
* Justin

## Minutes

### Effective ACR discovery

Henry: Issues related to optimizing the number of requests for effective ACR discovery:

> - [Support Trig serialization of Access Control Resources](https://github.com/solid/authorization-panel/issues/247)
> - [Define another relation for non-created ACR](https://github.com/solid/authorization-panel/issues/248)

Reviewer: ```suggestion
- [#247 — Support Trig serialization of Access Control Resources](https://github.com/solid/authorization-panel/issues/247)
- [#248 — Define another relation for non-created ACR](https://github.com/solid/authorization-panel/issues/248)
```

I describe my favorite option in issue #248, which adds a new "control" link type to link to the potential (not-yet-created) ACR for when a default ACR is being used.

> I need to create these issues to be able to link to them in order [PR 244: Effective ACR discovery](https://github.com/solid/authorization-panel/pull/244), and so for it to be merged.

Reviewer: ```suggestion
I need to create these issues to be able to link to them, in order for [PR 244: Effective ACR discovery](https://github.com/solid/authorization-panel/pull/244) to be merged.
```

Matthieu: Given that we have 4 potential solutions to minimising the number of requests involved in discovering the effective ACR, is everyone ok to merge #244?

Henry: Yes.

Elf: +1

Justin: +1

Matthieu: Merged.

Henry: The last solution might be useful for ACP too, if you want to weaken the requirement that every resource must have one ACR.

> Matthieu: I'm not sure that falls into the ACP model, but let's review once the draft ACP PR is ready for review (that will be done today).

Reviewer: ```suggestion
Matthieu: I'm not sure that falls into the ACP model, but let's review once the draft ACP PR is ready for review. That will be done today.
```

### Delegation use case

Henry: Could you give us a review of that use case, Elf?

Elf: How do you feel about using the [GNAP terminology](https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-00.html#name-roles) of Resource owner, server...?

Justin: I like using that terminology because I think it's well thought out and aligns with where we're going in terms of different actors and roles and responsibilities.

> Henry: Btw, a bit out of context, but how long is the GNAP spec since we mentioned it (I think it was 80 pages last time I looked)?

Reviewer: ```suggestion
Henry: Btw, a bit out of context, but how long is the GNAP spec since we mentioned it? I think it was 80 pages last time I looked...
```

The client cannot know how to continue.

But that is not the only use case. For example, we need a way for administrators to add rules on who can access (read and write).

But we also have a use case for [access by payment](https://solid.github.io/authorization-panel/authorization-ucr/#conditional-payment) or access for [users over 18](https://solid.github.io/authorization-panel/authorization-ucr/#uc-minimalcredentials). 
For those use cases, there need be no problem with the access control rules being known publically.  
In those cases, a `401` with a link to the rules that enable a client to know what to present to gain access (e.g., proof of being over 18) is completely fine.
Indeed, there are many real-life use-cases where it is widely accepted that access control rules should be public: 
 * If you walk in the countryside, you may see a "Do not trespass" sign. <img src="https://user-images.githubusercontent.com/124506/130351304-a5475264-ca59-4211-8435-e75a6628dc51.png" width="50%">
 * A shop has prices on items, that can be considered as access control rules: you need to present a money token to be able to take them out

>  *... 

Reviewer: ```suggestion
 * ... 
```
